fabtests/ubertest/ofi_atomic.h

```diff
@@ -61,21 +61,27 @@ typedef long double complex ofi_complex_long_double;
 #define ofi_atomic_isswap_op(op) \
 	(op >= OFI_SWAP_OP_START && op < OFI_SWAP_OP_LAST)
 
-extern void (*ofi_atomic_write_handlers[OFI_WRITE_OP_CNT][FI_DATATYPE_LAST])
+extern void (*ofi_atomic_write_handlers[OFI_WRITE_OP_CNT][OFI_DATATYPE_CNT])
 			(void *dst, const void *src, size_t cnt);
-extern void (*ofi_atomic_readwrite_handlers[OFI_READWRITE_OP_CNT][FI_DATATYPE_LAST])
+extern void (*ofi_atomic_readwrite_handlers[OFI_READWRITE_OP_CNT][OFI_DATATYPE_CNT])
 			(void *dst, const void *src, void *res, size_t cnt);
-extern void (*ofi_atomic_swap_handlers[OFI_SWAP_OP_CNT][FI_DATATYPE_LAST])
+extern void (*ofi_atomic_swap_handlers[OFI_SWAP_OP_CNT][OFI_DATATYPE_CNT])
 			(void *dst, const void *src, const void *cmp,
 			 void *res, size_t cnt);
 
-#define ofi_atomic_write_handler(op, datatype, dst, src, cnt) \
-	ofi_atomic_write_handlers[op][datatype](dst, src, cnt)
-#define ofi_atomic_readwrite_handler(op, datatype, dst, src, res, cnt) \
-	ofi_atomic_readwrite_handlers[op][datatype](dst, src, res, cnt)
-#define ofi_atomic_swap_handler(op, datatype, dst, src, cmp, res, cnt) \
-	ofi_atomic_swap_handlers[op - OFI_SWAP_OP_START][datatype](dst, src, \
-								cmp, res, cnt)
+#define ofi_atomic_write_handler(op, datatype)				\
+	ofi_atomic_write_handlers[op][datatype]
+#define ofi_atomic_readwrite_handler(op, datatype)			\
+	ofi_atomic_readwrite_handlers[op][datatype]
+#define ofi_atomic_swap_handler(op, datatype)				\
+	ofi_atomic_swap_handlers[op - OFI_SWAP_OP_START][datatype]
+
+#define ofi_atomic_write_op(op, datatype, dst, src, cnt)		\
+	ofi_atomic_write_handler(op, datatype)(dst, src, cnt)
+#define ofi_atomic_readwrite_op(op, datatype, dst, src, res, cnt)	\
+	ofi_atomic_readwrite_handler(op, datatype)(dst, src, res, cnt)
+#define ofi_atomic_swap_op(op, datatype, dst, src, cmp, res, cnt)	\
+	ofi_atomic_swap_handler(op, datatype)(dst, src, cmp, res, cnt)
```

Comment:
Help me understand the double indirection here -- adding the 'op' macros to call the 'handler' macros.

Reply:
If the compiler doesn't have __int128 support, then then the handlers are going to be NULL; this is assumed not to be the case for any other types. I needed to modify verify_atomic() in verify.c to test whether the handler existed and I wanted a single macro to give me the pointer to the handler. The original xxx_handler() macro naming made a lot more sense for that.

This would only happen in the case when the provider supported 128-bit ints, but the compiler did not. So, I also added a check for this case to just report "UNVERIFIED" which would mean the 128-bit op ran without error, but we don't have the logic to test it.